phpBB/phpbb/db/extractor/factory.php

```diff
@@ -0,0 +1,84 @@
+<?php
+/**
+*
+* This file is part of the phpBB Forum Software package.
+*
+* @copyright (c) phpBB Limited <https://www.phpbb.com>
+* @license GNU General Public License, version 2 (GPL-2.0)
+*
+* For full copyright and license information, please see
+* the docs/CREDITS.txt file.
+*
+*/
+
+namespace phpbb\db\extractor;
+
+/**
+* A factory which serves the suitable extractor instance for the given dbal
+*/
+class factory
+{
+	/**
+	 * @var string
+	 */
+	private $phpbb_root_path;
+
+	/**
+	 * @var \phpbb\request\request_interface
+	 */
+	private $request;
+
+	/**
+	* Extractor factory constructor
+	*
+	* @param \phpbb\request\request_interface	$request
+	* @param string	$phpbb_root_path
+	*/
+	public function __construct($request, $phpbb_root_path)
+	{
+		$this->request			= $request;
+		$this->phpbb_root_path	= $phpbb_root_path;
+	}
+
+	/**
+	* DB extractor factory getter
+	*
+	* @param \phpbb\db\driver\driver_interface	$db_driver
+	*/
+	public function get($db_driver)
+	{
+		// Return the appropriate DB extractor
+		switch ($db_driver->get_sql_layer())
+		{
+			case 'mysqli':
+			case 'mysql4':
+			case 'mysql':
+				return new \phpbb\db\extractor\mysql_extractor($this->phpbb_root_path, $this->request, $db_driver);
```

Comment:
Shouldn't we maybe use a service collection in the factory instead of creating a new object every time?